### PR TITLE
[NEW][Enterprise] Micro services

### PR DESCRIPTION
**Pull request by sampaiodiego**
## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves a current function)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Hotfix (a major bugfix that has to be merged asap)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat)
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/lib/server/functions/setRoomAvatar.js [hunk 1]
```diff
@@ -13,7 +14,7 @@ export const setRoomAvatar = function(rid, dataURI, user) {
 	if (!dataURI) {
 		fileStore.deleteByRoomId(rid);
 		Messages.createRoomSettingsChangedWithTypeRoomIdMessageAndUser('room_changed_avatar', rid, '', user);
-		Notifications.notifyLogged('updateAvatar', { rid });
+		notifications.notifyLogged('updateAvatar', { rid });
```
Comment (sampaiodiego) [inline app/lib/server/functions/setRoomAvatar.js:17]:
```suggestion
		api.broadcast('room.avatarUpdate', { rid });
```

File: app/lib/server/lib/msgStream.js [hunk 2]
```diff
@@ -1,57 +1,3 @@
...
-					eventName,
-					args: [args, options],
-				}));
-			}
-		});
-	}
-}
-
-export const msgStream = new MessageStream('room-messages');
+export const msgStream = notifications.streamRoomMessage;
```
Comment (sampaiodiego) [inline app/lib/server/lib/msgStream.js:3]: this doesn't look to be used anywhere

File: imports/users-presence/server/activeUsers.js [hunk 3]
```diff
@@ -13,19 +13,14 @@ export const STATUS_MAP = {
 
 export const setUserStatus = (user, status/* , statusConnection*/) => {
 	const {
-		_id,
+		_id: uid,
```
Comment (sampaiodiego) [inline imports/users-presence/server/activeUsers.js:16]:
```suggestion
		_id,
```

File: imports/users-presence/server/activeUsers.js [hunk 4]
```diff
@@ -13,19 +13,14 @@ export const STATUS_MAP = {
...
 
 	// since this callback can be called by only one instance in the cluster
 	// we need to broadcast the change to all instances
-	Notifications.notifyLogged('user-status', [
-		_id,
-		username,
-		STATUS_MAP[status],
-		statusText,
-	]);
+	api.broadcast('userpresence', { user: { status, _id: uid, username, statusText } }); // remove username
```
Comment (sampaiodiego) [inline imports/users-presence/server/activeUsers.js:23]:
```suggestion
	api.broadcast('userpresence', { user: { status, _id, username, statusText } }); // remove username
```

File: package.json [hunk 5]
```diff
@@ -217,9 +221,12 @@
 		"mime-db": "^1.44.0",
 		"mime-type": "^3.1.0",
 		"mkdirp": "^0.5.5",
+		"moleculer": "^0.14.10",
 		"moment": "^2.27.0",
 		"moment-timezone": "^0.5.31",
 		"mongodb": "^3.6.0",
+		"msgpack5": "^4.2.1",
```
Comment (sampaiodiego) [inline package.json:228]: do we need `msgpack5`?

File: server/sdk/index.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import { IEnterpriseSettings } from './types/IEnterpriseSettings';
+
+// TODO think in a way to not have to pass the service name to proxify here as well
+export const Authorization = proxifyWithWait<IAuthorization>('authorization');
+export const Presence = proxifyWithWait<IPresence>('presence');
+export const Account = proxifyWithWait<IAccount>('accounts');
+export const License = proxifyWithWait<ILicense>('license');
+export const MeteorService = proxifyWithWait<IMeteor>('meteor');
+
+// Calls without wait. Means that the service is optional and the result may be ane error
```
Comment (sampaiodiego) [inline server/sdk/index.ts:19]:
```suggestion
// Calls without wait. Means that the service is optional and the result may be an error
```

File: server/sdk/lib/Api.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	registerService(instance: ServiceClass): void {
+		this.services.add(instance);
+
+		if (this.broker) {
+			this.broker.createService(instance);
+		}
+	}
+
+	async call(method: string, data: any): Promise<any> {
+		// console.log('api call', method, this.broker.constructor.name);
```
Comment (sampaiodiego) [inline server/sdk/lib/Api.ts:40]:
```suggestion
```

File: server/sdk/lib/Api.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+		return this.broker.call(method, data);
+	}
+
+	async waitAndCall(method: string, data: any): Promise<any> {
+		// console.log('api call', method, this.broker.constructor.name);
```
Comment (sampaiodiego) [inline server/sdk/lib/Api.ts:45]:
```suggestion
```

File: app/lib/server/lib/msgStream.js (hunk 2 above)
Comment (sampaiodiego) [reply app/lib/server/lib/msgStream.js:3]: it is :( need to change that then

File: package.json (hunk 5 above)
Comment (sampaiodiego) [reply package.json]: no, we don't.. removed =)

**Review by rodrigok — APPROVED** [review]

**Pull request merged**